### ISSUE TITLE
What's wrong with this output?

### ISSUE THREAD

**Issue by A-new-b** [TencentARC/SmartEdit#9]
First of all, thank you for your work. 
However, after a long period of debugging and error correction, I believe there are problems with the inference code of the model.
I used your [dataset](https://drive.google.com/drive/folders/1SMkQe1U9av4YNML5wqOLN7crLiNs0aTF)'s "5-Multiple-Objects 001.png" as test sample. 
![001](https://github.com/TencentARC/SmartEdit/assets/43344034/3a12da17-76d2-4e2c-bcb9-83fbd2f7833c)

Instruction:
**Change the left goat to a horse  CLIP: horse**

run this as followed:

`python test/DS_SmartEdit_test.py --is_understanding_scenes True --model_name_or_path "./checkpoints/vicuna-7b-v1.1" --LLaVA_model_path "checkpoints/LLaVA-7b-v1.1" --save_dir 'datasets/results1' --steps 15000 --total_dir "./checkpoints/SmartEdit-7B" --sd_qformer_version "v1.1-7b" --resize_resolution 256 --ReasonEdit_benckmark_dir 'datasets/ReasonEdit2'`

and got this as results:
![0001_T7 5_I1 4](https://github.com/TencentARC/SmartEdit/assets/43344034/cecca1bf-c3e7-47d2-8f8f-8633fa426f2f)
![0001_T7 5_I1 5](https://github.com/TencentARC/SmartEdit/assets/43344034/ada3b632-7d01-4ecc-9dcf-b2e0da04f6a0)
![0001_T7 5_I1 6](https://github.com/TencentARC/SmartEdit/assets/43344034/437b0e16-7817-4f7e-809e-d6d9a96a48e1)

![0001_T7 5_I1 5](https://github.com/TencentARC/SmartEdit/assets/43344034/9adc2346-0397-46e3-a0b2-8caaf6969709)

**Comment by A-new-b** [comment]
And I'd like to share the modifications what I made. 
First, I removed the code [here](https://github.com/TencentARC/SmartEdit/blob/47dc8f0c09ad1e9e3f3c0b2fcbee03bfb9bdc472/test/DS_SmartEdit_test.py#L239C1-L240C77) for I thought this
`model_.model = PeftModel_for_LLM.from_pretrained(model_.model, LLM_sub_dir, adapter_name=adapter_name)` had already loaded the lora ckpt for llava. 
And I tried `self.model.load_state_dict(LLM_ckpt_new, strict=False)` [here](https://github.com/TencentARC/SmartEdit/blob/47dc8f0c09ad1e9e3f3c0b2fcbee03bfb9bdc472/model/DS_SmartEdit_model.py#L316). the results were the same.
[this code](https://github.com/TencentARC/SmartEdit/blob/47dc8f0c09ad1e9e3f3c0b2fcbee03bfb9bdc472/test/DS_SmartEdit_test.py#L119) model_.config.num_new_tokens = 35 , obviously these couldn't be equal `lm_img_token_states.shape[1] == num_new_tokens`, I just use model_.config.num_new_tokens-3.
What's more, I feel that the hidden statues in train process and test process is not the same.

**Comment by yuzhou914** [comment]
Thanks for your interest. It seems that the decoding process from LLM gets wrong, and it is not normal. The correct decoding process from LLM will get correct decoding results, that is you will get 32003~32034 tokens, the correct order of 32 tokens. I don't know if your llava checkpoint is prepared correctly, if you are not sure, please test the llava conversation first (if the conversation is correct or not).

**Comment by A-new-b** [comment]
> Thanks for your interest. It seems that the decoding process from LLM gets wrong, and it is not normal. The correct decoding process from LLM will get correct decoding results, that is you will get 32003~32034 tokens, the correct order of 32 tokens. I don't know if your llava checkpoint is prepared correctly, if you are not sure, please test the llava conversation first (if the conversation is correct or not).

I have alreay tested the llava. its output is right. but I still got the result like [this issue](https://github.com/TencentARC/SmartEdit/issues/8)

**Comment by A-new-b** [comment]
Are you saying that I should get one copy of the embedding for each of the <img1>.....to <img32> tokens [here](https://github.com/TencentARC/SmartEdit/blob/47dc8f0c09ad1e9e3f3c0b2fcbee03bfb9bdc472/test/DS_SmartEdit_test.py#L122C5-L124C66) in inference_llm and concatenate them, instead of getting them in a loop?

**Comment by yuzhou914** [comment]
1. Firstly, you only need to merge lora and llava weights and do not need to make any further change for the code, it must be correct, the correct decoding process will automatically get 32003~32034 tokens. 2. As you may notice the issue, the preparation of llava is incorrect, the llama delta weight is under policy, you need to firstly convert the delta weight to the correct llava weight, please carefully read the llava instruction.

btw, could you please show your llava conversation, for example, one test image and text, and the answer by llava.

**Comment by A-new-b** [comment]
I am sure that my llava model is correct. I have convert the delta weight to the correct llava weight by
`python3 -m llava.model.apply_delta \
    --base /path/to/llama-7b \
    --target /output/path/to/LLaVA-7B-v1.1 \
    --delta checkpoints/LLaVA-Lightning-7B-delta-v1-1`
and there is the test example.
**images**:
![COCO_train2014_000000360306](https://github.com/TencentARC/SmartEdit/assets/43344034/c0c6d340-0386-410c-bc0e-046e43357a53)
**code**:
```python
import argparse
import torch

from llava.constants import IMAGE_TOKEN_INDEX, DEFAULT_IMAGE_TOKEN, DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN
from llava.conversation import conv_templates, SeparatorStyle
from llava.model.builder import load_pretrained_model
from llava.utils import disable_torch_init
from llava.mm_utils import process_images, tokenizer_image_token, get_model_name_from_path, KeywordsStoppingCriteria

from PIL import Image

import requests
from PIL import Image
from io import BytesIO
from transformers import TextStreamer
import os
from tqdm import tqdm

import json
import re


def load_image(image_file):
    if image_file.startswith('http://') or image_file.startswith('https://'):
        response = requests.get(image_file)
        image = Image.open(BytesIO(response.content)).convert('RGB')
    else:
        image = Image.open(image_file).convert('RGB')
    return image

def main(args):
    # Model
    disable_torch_init()

    model_name = get_model_name_from_path(args.model_path)

    if "llama-2" in model_name.lower():
        conv_mode = "llava_llama_2"
    elif "mistral" in model_name.lower():
        conv_mode = "mistral_instruct"
    elif "v1.6-34b" in model_name.lower():
        conv_mode = "chatml_direct"
    elif "v1" in model_name.lower():
        conv_mode = "llava_v1"
    elif "mpt" in model_name.lower():
        conv_mode = "mpt"
    else:
        conv_mode = "llava_v0"

    if args.conv_mode is not None and conv_mode != args.conv_mode:
        print('[WARNING] the auto inferred conversation mode is {}, while `--conv-mode` is {}, using {}'.format(conv_mode, args.conv_mode, args.conv_mode))
    else:
        args.conv_mode = conv_mode
    tokenizer, model, image_processor, context_len = load_pretrained_model(args.model_path, args.model_base, model_name, args.load_8bit, args.load_4bit)


    image_files = [load_image(args.image_path)]

    question_list = [
        "Please describe this picture. "
    ]
    for image_file in tqdm(image_files):
        conv = conv_templates[args.conv_mode].copy()
        
        image = image_file
        image_size = image.size
        # Similar operation in model_worker.py
        image_tensor = process_images([image], image_processor, model.config)
        if type(image_tensor) is list:
            image_tensor = [image.to(model.device, dtype=torch.float16) for image in image_tensor]
        else:
            image_tensor = image_tensor.to(model.device, dtype=torch.float16)

        inp = question_list[0]

        if image is not None:
            # first message
            if model.config.mm_use_im_start_end:
                inp = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + '\n' + inp
            else:
                inp = DEFAULT_IMAGE_TOKEN + '\n' + inp
            conv.append_message(conv.roles[0], inp)
            image = None
        else:
            # later messages
            conv.append_message(conv.roles[0], inp)
        conv.append_message(conv.roles[1], None)
        prompt = conv.get_prompt()
        input_ids = tokenizer_image_token(prompt, tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').unsqueeze(0).cuda()

        stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
        keywords = [stop_str]
        stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)

        with torch.inference_mode():
            output_ids = model.generate(
                input_ids,
                images=image_tensor,
                do_sample=True,
                temperature=0.2,
                max_new_tokens=1024,
                use_cache=True,
                stopping_criteria=[stopping_criteria])

        input_token_len = input_ids.shape[1]
        n_diff_input_output = (input_ids != output_ids[:, :input_token_len]).sum().item()
        if n_diff_input_output > 0:
            print(f'[Warning] {n_diff_input_output} output_ids are not the same as the input_ids')
        outputs = tokenizer.batch_decode(output_ids[:, input_token_len:], skip_special_tokens=True)[0]
        outputs = outputs.strip()
        if outputs.endswith(stop_str):
            outputs = outputs[:-len(stop_str)]
        outputs = outputs.strip()
        print(outputs)
        





if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--model-path", type=str, default="/path/LLaVA-7b-v1.1")
    parser.add_argument("--model-base", type=str, default=None)
    parser.add_argument("--image-file", type=str, required=False)
    parser.add_argument("--device", type=str, default="cuda")
    parser.add_argument("--conv-mode", type=str, default=None)
    parser.add_argument("--temperature", type=float, default=0.2)
    parser.add_argument("--max-new-tokens", type=int, default=512)
    parser.add_argument("--load-8bit", action="store_true")
    parser.add_argument("--load-4bit", action="store_true")
    parser.add_argument("--image-path", type=str,default="/path/COCO_train2014_000000360306.jpg")
    parser.add_argument("--labels",default="happiness, sadness, anger, fear, disgust, surprise, neutral", type=str)
    parser.add_argument("--debug", action="store_true")
    args = parser.parse_args()
    main(args)
```

**output**：
The picture shows a man standing in a kitchen, with a dining table and chairs nearby. There are various items on the table, including a plate, bowl, and a bottle. The man appears to be preparing to eat, and there is a dog in the scene as well. The overall atmosphere of the image is casual and domestic, with a focus on the man's daily routine and the comfort of his home environment.

**Comment by A-new-b** [comment]
What's more have you notice that [test code](https://github.com/TencentARC/SmartEdit/blob/47dc8f0c09ad1e9e3f3c0b2fcbee03bfb9bdc472/test/DS_SmartEdit_test.py#L54C4-L78C36) inference llm for 512 times. but the [train code](https://github.com/TencentARC/SmartEdit/blob/47dc8f0c09ad1e9e3f3c0b2fcbee03bfb9bdc472/model/DS_SmartEdit_model.py#L625C1-L627C70) get the hidden status at once inference. I think that is the key to these problem.

**Comment by yuzhou914** [comment]
Thanks for your sharing, and I will inference and find out the problem. The inference code is correct since when training, the LLM learns to output the 32 new learnable tokens, which means only these 32 tokens are not masked. While when inference, the LLM aims to decode the 32 new tokens only, which stands for the aggregation information of input image and instruction. If the inference is correct, the 32 tokens will be received in correct order, and the "for" loop will break, it will not last 512 times.

I will re-execute the inference code, and try to solve this problem in these two days, I still guess there might be some mistakes for LLM weights, the inference code is tested before upload.

**Comment by A-new-b** [comment]
> Thanks for your sharing, and I will inference and find out the problem. The inference code is correct since when training, the LLM learns to output the 32 new learnable tokens, which means only these 32 tokens are not masked. While when inference, the LLM aims to decode the 32 new tokens only, which stands for the aggregation information of input image and instruction. If the inference is correct, the 32 tokens will be received in correct order, and the "for" loop will break, it will not last 512 times.
> 
> I will re-execute the inference code, and try to solve this problem in these two days, I still guess there might be some mistakes for LLM weights, the inference code is tested before upload.

Okay, thanks for your reply. I hope you can resolve this issue as soon as possible.

**Comment by yuzhou914** [comment]
Hi, we have updated how to solve, please take a look.

**Comment by A-new-b** [comment]
Thanks for your reminder. Testing after updating the model indeed yielded satisfactory results. However, I'd like to point out that the images in the dataset are in JPG format, not PNG.

**Comment by liuxiaoyu1104** [comment]
When I convert the delta weight to the correct llava weight by
python3 -m llava.model.apply_delta \ --base /path/to/llama-7b \ --target /output/path/to/LLaVA-7B-v1.1 \ --delta checkpoints/LLaVA-Lightning-7B-delta-v1-1,  I encountered the following error.
![2355096287](https://github.com/TencentARC/SmartEdit/assets/58921114/e1f6cb22-7c75-44c1-bd38-f8c84b491498)

Could you please provide me with the specific path of "/path/to/llama-7b" @A-new-b
